crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use ruma::{
+    events::{
+        poll::{start::PollStartEventContent, unstable_start::UnstablePollStartEventContent},
+        receipt::{ReceiptThread, ReceiptType},
+        room::message::Relation,
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent, OriginalSyncMessageLikeEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, OwnedEventId, RoomId, UserId,
+};
+use tracing::{field::display, instrument, trace};
+
+use super::BaseClient;
+use crate::{error::Result, store::StateChanges, RoomInfo};
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    // TODO: return a reference or some kind of iterator
+    fn for_room(&self, room_id: &RoomId) -> Vec<SyncTimelineEvent>;
+}
+
+impl PreviousEventsProvider for () {
+    fn for_room(&self, _: &RoomId) -> Vec<SyncTimelineEvent> {
+        Vec::new()
+    }
+}
+
+#[instrument(skip_all, fields(room_id))]
+pub(crate) async fn compute_notifications(
```

Comment:
Can we rename `compute_notifications` to `compute_notifications_for_encrypted_room`? It makes things super explicit and I prefer that.

Reply:
No, it's used also for unencrypted rooms in the latest version. (It contains more info than `UnreadNotificationsCount`, so we can't use that one.)